pkg/controller/util/endpoint/controller_utils.go

```diff
@@ -32,6 +34,73 @@ import (
 // EndpointsMatch is a type of function that returns true if pod endpoints match.
 type EndpointsMatch func(*v1.Pod, *v1.Pod) bool
 
+// ServiceSelectorCache is a cache of service selectors to avoid high CPU consumption caused by frequent calls to AsSelectorPreValidated (see #73527)
+type ServiceSelectorCache struct {
+	cache map[string]labels.Selector
+	lock  sync.Mutex
+}
+
+// NewServiceSelectorCache init ServiceSelectorCache for both endpoint controller and endpointSlice controller.
+func NewServiceSelectorCache() *ServiceSelectorCache {
+	return &ServiceSelectorCache{
+		cache: map[string]labels.Selector{},
+	}
+}
+
+// Get return selector and existence in ServiceSelectorCache by key.
+func (sc *ServiceSelectorCache) Get(key string) (labels.Selector, bool) {
+	selector, ok := sc.cache[key]
```

Comment:
I wonder why there is no lock taken.
Wouldn't this introduce race when there is concurrent write (line 61 below) ?

Reply:
Yes, I think read write lock maybe better.